doxygen/contributor_help_pages/getting_started.md

```diff
@@ -57,84 +60,136 @@ Within each of those folders you will find any one of the following folders
+You can contribute a new function by writing the function in the `prim/fun` folder. Though the function will need to accept arithmetic, reverse mode (@ref stan::math::var  ), and forward mode (`fvar<T>`) scalar types and should be tested using the `expect_ad()` testing framework. The `rev`, `fwd`, `mix`, and `opencl` folders are used for adding specialization for:
 
 1. Performance
-2. Better numerical behavior
+2. Better numerical behavior of gradient calculations
 4. Utilize particular hardware
 
 ## Adding a Simple Example function
 
-A generic function that performs a dot product on itself could be written in `prim/fun` as
+A generic function that takes in an Eigen vector and performs a dot product on itself could be written in `prim/fun` as
 
 ```cpp
 // stan/math/prim/fun/dot_self.hpp
-template <typename EigVec>
-inline double dot_self(const EigVec& x) {
-  const auto& x_ref = to_ref(x); // (1)
-  value_type_t<EigVec> sum_x = 0.0; // (2)
+template <typename EigVec, require_eigen_vector_t<EigVec>* = nullptr> // (1)
+inline value_type_t<EigVec> dot_self(const EigVec& x) {
+  const auto& x_ref = to_ref(x); // (2)
+  value_type_t<EigVec> sum_x = 0.0; // (3)
   for (int i = 0; i < x.size(); ++i) {
-    sum_x += x_ref.coeff(i) * x_ref.coeff(i); // (3)
+    sum_x += x_ref.coeff(i) * x_ref.coeff(i); // (4)
   }
   return sum_x;
 }
 ```
 
-This is pretty standard C++ besides (1), (2), and (3) which I'll go over here.
+Do keep in mind that in reality we would just want to return `x.squaredNorm()` on the input vector `x`. But unrolling this `dot_self()` example gives us a nice example with a few gotchas we need to watch for.
+
+There are several odd parts to this function labeled (1), (2), (3), and (4) which I'll go over below.
+
+#### (1) Using Stan's require type traits to specialize overloads.
+
+The Stan math library requires that all functions are able to accept Eigen expression templates, hence the general `EigVec` template above. Though we want general templates parameters to the functions, we also want to ensure that only a restricted subset of types are able to be accepted for a particular function. For example, without the require in the above function it would be perfectly valid to pass in an `Eigen::MatrixXd`, but that would not be good! In order to restrict the types that can go into a function the math library uses the `require` type traits to detecting which function a particular type should use. This is very similar conceptually to [C++20 requires](https://en.cppreference.com/w/cpp/language/constraints#Constraints).
+
+To avoid document duplication I recommend you skim over @ref require_meta_doc The guide there explains the different styles and patterns of requires that we use throughout the math library as well as in this example. At the bottom of that page you will see subsections which will show you the `requires` we have already setup for types we commonly see.
```

Comment:
> Missing period at end of first sentence. 

Fixed

> I don't get the comment about "document duplication", though---what does that mean?

I just mean that I'd like to avoid explaining something in two different places so if we ever have to update it then we only need to update it in once place